### PR TITLE
[v0.1] align APIs for vision/ImagePredictor/ObjectDetector

### PR DESCRIPTION
**Pull request by zhreshold**
*Issue #, if available:*

*Description of changes:*
- API adjustments for vision part, to align with TabularPredictor
- GPU memory warning for too large custom batch sizes

Update: re-ran sanity fitting batch test with all d8 datasets over the weekend, all passed with reasonable validation scores for all presets with `epochs=1` 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: vision/src/autogluon/vision/image_prediction/image_prediction.py
Comment (Innixma) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:275]: recommend it just be called `data`, since this isn't necessarily used for tuning.

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (Innixma) [inline vision/src/autogluon/vision/object_detection/object_detection.py:45]: same comments on num_cpus and num_gpus as in ImagePredictor

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (Innixma) [inline vision/src/autogluon/vision/object_detection/object_detection.py:41]: Same comment regarding hyperparameter_tune_kwargs

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (Innixma) [inline vision/src/autogluon/vision/object_detection/object_detection.py:105]: It might be nicer if downstream None was handled instead of setting extremely high values. Not necessary for v0.1, more an observation.

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (Innixma) [inline vision/src/autogluon/vision/object_detection/object_detection.py:166]: ditto

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (Innixma) [inline vision/src/autogluon/vision/object_detection/object_detection.py:29]: will we ever have a reason to have different eval_metric or problem_type for this?

**Review by Innixma — COMMENTED** [review]
Looks good, just a few minor tweaks to simplify the hyperparameter tuning arguments

**Comment by szha** [comment]
Job PR-882-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/2/index.html

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (Innixma) [reply vision/src/autogluon/vision/object_detection/object_detection.py:105]: Also, why is config specifying 'time_limits'? Can we make it 'time_limit'?

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (jwmueller) [reply vision/src/autogluon/vision/object_detection/object_detection.py:29]: seems important to allow for other eval_metric in the future

File: vision/src/autogluon/vision/image_prediction/image_prediction.py
Comment (jwmueller) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:69]: "reserved for tuning processes" -> "reserved for model selection and hyperparameter-tuning"

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (jwmueller) [inline vision/src/autogluon/vision/object_detection/object_detection.py:57]: "Holdout tuning data for validation" -> "Held-out validation data reserved for model selection and hyperparameter-tuning"

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (jwmueller) [inline vision/src/autogluon/vision/object_detection/object_detection.py:63]: "shuffling" -> "shuffling data"

File: vision/src/autogluon/vision/image_prediction/image_prediction.py
Comment (jwmueller) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:76]: "shuffling" -> "shuffling data"

File: vision/src/autogluon/vision/image_prediction/image_prediction.py
Comment (jwmueller) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:174]: "your batch_size setting" -> "whether your data mini-batches will fit on the GPU for the specified batch_size"

File: vision/src/autogluon/vision/image_prediction/image_prediction.py
Comment (jwmueller) [inline vision/src/autogluon/vision/image_prediction/image_prediction.py:178]: "batch-size" -> "batch_size"

Comment (jwmueller) [reply vision/src/autogluon/vision/image_prediction/image_prediction.py:178]: add to warning: "Try reducing "batch_size" if you encounter memory issues"

File: vision/src/autogluon/vision/object_detection/object_detection.py
Comment (zhreshold) [reply vision/src/autogluon/vision/object_detection/object_detection.py:105]: this internal naming and behavior change will need to be addressed in gluoncv and it may not be necessary for 0.1 release

**Comment by szha** [comment]
Job PR-882-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/6/index.html

File: vision/src/autogluon/vision/detector/detector.py
Comment (Innixma) [inline vision/src/autogluon/vision/detector/detector.py:165]: consider using `from autogluon.core.scheduler.scheduler_factory import scheduler_factory` to simplify this code and align with Tabular.

Comment (Innixma) [inline vision/src/autogluon/vision/detector/detector.py:165]: Perhaps sanity check that no unknown kwargs were passed.

File: vision/src/autogluon/vision/detector/detector.py
Comment (Innixma) [inline vision/src/autogluon/vision/detector/detector.py:185]: This does not print out the value of max_reward. Missing `{}`. Also, the if statement doesn't check for `> 1` as the error states.

File: vision/src/autogluon/vision/predictor/predictor.py
Comment (Innixma) [inline vision/src/autogluon/vision/predictor/predictor.py:94]: same comment as in ObjectDetection

File: vision/src/autogluon/vision/predictor/predictor.py
Comment (Innixma) [inline vision/src/autogluon/vision/predictor/predictor.py:246]: Ditto comment: Missing `{}` to print actual value

File: vision/src/autogluon/vision/predictor/predictor.py
Comment (Innixma) [inline vision/src/autogluon/vision/predictor/predictor.py:226]: ditto same comment: Consider using `scheduler_factory`

File: vision/src/autogluon/vision/predictor/predictor.py
Comment (Innixma) [inline vision/src/autogluon/vision/predictor/predictor.py:375]: `file_name` -> `path`, ideally it should be the same path value as specified by user or available via `predictor.path`.

File: vision/src/autogluon/vision/detector/detector.py
Comment (zhreshold) [reply vision/src/autogluon/vision/detector/detector.py:77]: introducing another `resource` concept is adding additional complexity IMO, however, the value `nthreads_per_trial` or `num_cpus` is useful even with HPO turned off.

File: vision/src/autogluon/vision/detector/detector.py
Comment (zhreshold) [reply vision/src/autogluon/vision/detector/detector.py:83]: the logic is to stop HPO when either `time_limit` or `num_trials` reaches the limit, whichever comes first. What do you suggest for the docstring?

File: vision/src/autogluon/vision/detector/detector.py
Comment (Innixma) [reply vision/src/autogluon/vision/detector/detector.py:77]: Why is it valuable if HPO is turned off? That isn't what is communicated by the documentation:

```
Number of CPU threads for each trial, if `None`, will detect the # cores on current instance.
```

Are you referring to the single model training if HPO is turned off? If that is the case, shouldn't it be specified in the model hyperparameters?

File: vision/src/autogluon/vision/detector/detector.py
Comment (zhreshold) [reply vision/src/autogluon/vision/detector/detector.py:88]: Yes, theoretically it's almost impossible to reach 0.9 mean average precision for bonding boxes. Even if a model does, the prediction difference between 0.9 mAP and 1.0 mAP is neglected. 

My input of 0.9 here is to avoid the searcher desperately exploring the best configuration even if the result is already saturated(good for the average users, bad for competition?) What I can do is to put 1.0 back as default, or print a message to tell if the `max_reward` is met but you can possibly improve it.

@Innixma What's you suggestion?

File: vision/src/autogluon/vision/detector/detector.py
Comment (Innixma) [reply vision/src/autogluon/vision/detector/detector.py:83]:
```
The maximum amount of HPO trials to perform, without going over the specified `time_limit`.
```

Additionally, I'd recommend adding a link to FIFOScheduler documentation in the overall hyperparameter_tune_kwargs docs.

```
For valid dictionary keys, refer to :class:`autogluon.core.scheduler.FIFOScheduler` documentation.
    The 'searcher' key is required when providing a dict.
```

Example, scroll to kwargs hyperparameter_tune_kwargs to see link in doc: http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-889/8/api/autogluon.task.html#autogluon.tabular.TabularPredictor.fit

File: vision/src/autogluon/vision/detector/detector.py
Comment (zhreshold) [reply vision/src/autogluon/vision/detector/detector.py:88]: Adding `presets` though, where `best_quality` mode will override `max_reward` to 1 though

Comment (Innixma) [reply vision/src/autogluon/vision/detector/detector.py:88]: I'll leave this to your decision since you are more familiar with what 0.9 mAP means, but I would suggest that so long as the user specified a long enough time_limit, there is no reason to stop training when improvements in score could still be gained. Perhaps with a higher quality preset it can be set to 1?

File: vision/src/autogluon/vision/predictor/predictor.py
Comment (Innixma) [inline vision/src/autogluon/vision/predictor/predictor.py:105]: I feel that this should default to 1 or effectively not be set. I think there is little reason to stop HPO early when achieving 95% accuracy, especially for datasets that are imbalanced and it is trivial to get 95% accuracy without much meaningful benefits.

Comment (Innixma) [reply vision/src/autogluon/vision/predictor/predictor.py:105]: Another reason for not setting this at all would be if in future we had `log_loss` as an evaluation metric. A max_reward of 0.95 wouldn't make sense in this case.

File: vision/src/autogluon/vision/configs/presets_configs.py
Comment (Innixma) [inline vision/src/autogluon/vision/configs/presets_configs.py:22]: This will end very poorly if someone does not specify a time_limit. We should ensure that a time limit is set for this, otherwise it will run 1024 trials.

Comment (Innixma) [reply vision/src/autogluon/vision/configs/presets_configs.py:22]: Same comments for all of the presets. Any time num_trials is specified automatically, we probably want to ensure time_limit is specified by user

File: vision/src/autogluon/vision/configs/presets_configs.py
Comment (Innixma) [inline vision/src/autogluon/vision/configs/presets_configs.py:64]: Why 0.85? This seems extremely low. I generally think we should avoid any specification of max_reward, since an acceptable accuracy differs drastically between problems.

File: vision/src/autogluon/vision/configs/presets_configs.py
Comment (Innixma) [inline vision/src/autogluon/vision/configs/presets_configs.py:86]: same comments here as in image predictor presets

File: vision/src/autogluon/vision/configs/presets_configs.py
Comment (Innixma) [inline vision/src/autogluon/vision/configs/presets_configs.py:161]: Do you think tabular's preset functionality is general enough to move to core so we don't have to copy paste these functions?

Comment (zhreshold) [reply vision/src/autogluon/vision/configs/presets_configs.py:161]: moving to core is nice, see my modifications, I prefer a named tuple or mapping so that I won't need to import specific preset dict from ag.core

File: vision/src/autogluon/vision/configs/presets_configs.py
Comment (zhreshold) [reply vision/src/autogluon/vision/configs/presets_configs.py:22]: here's my confusing part, if running things without a wall clock time limit is scary, should we at least safeguard the default time_limit with a reasonable value, say 1 day or 3 days?

Comment (Innixma) [reply vision/src/autogluon/vision/configs/presets_configs.py:22]: Perhaps we can set default to 4 hours in the presets and warn the user if they did not set it themselves. Whatever solution we go with, we need to ensure that the user doesn't unknowingly start a job by specifying a preset that will take weeks+ to finish when they never touched time_limit.

I would personally be in favor of raising an exception if the user specified a preset but did not specify time_limit, or otherwise set num_trials=1 if they only specified a preset but no time limit. Alternatively, we could set time_limit default value in the preset itself (and user can freely override if they wish).

Example:

```
    best_quality={
        'hyperparameters': {
            'model': Categorical('resnet50_v1b', 'resnet101_v1d', 'resnest200'),
            'lr': Real(1e-2, 1e-6, log=True),
            'batch_size': Categorical(8, 16, 32, 64, 128),
            'epochs': 200
            },
        'hyperparameter_tune_kwargs': {
            'num_trials': 1024,
            'search_strategy': 'bayesopt',
            'max_reward': 1.0
        },
        'time_limit': 86400,
    },
```

**Comment by szha** [comment]
Job PR-882-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/9/index.html

**Comment by szha** [comment]
Job PR-882-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/10/index.html

File: mxnet/src/autogluon/mxnet/task/classifier.py
Comment (jwmueller) [reply mxnet/src/autogluon/mxnet/task/classifier.py:36]: All the documentation should drop the usage of the term `task` since ImagePrediction will be dropped and ImagePredictor is not a task.

File: vision/src/autogluon/vision/predictor/predictor.py
Comment (jwmueller) [reply vision/src/autogluon/vision/predictor/predictor.py]: Agreed I think this should default to infinity and we should not assume metrics will always lie in [0,1].

File: vision/src/autogluon/vision/configs/presets_configs.py
Comment (jwmueller) [reply vision/src/autogluon/vision/configs/presets_configs.py:22]: Alternative suggestion:

- Do not set time-limits by default (since users expect things to take longer for larger datasets anyway).

- Set num_trials default (probably much lower than 1024 though. Perhaps for the different presets we could have: num_trials = 1 (for medium), 10 (for good), 100 (for best).

- Whenever user specifies time_limit, this overrides num_trials. So we don't run for min{time_limit, num_trials}, but instead just run for whatever time_limit the user specified (which may involve more trials than the default num_trials). If user specifies both time_limit and num_trials, then throw Exception.

**Comment by szha** [comment]
Job PR-882-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/12/index.html

**Comment by szha** [comment]
Job PR-882-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/15/index.html

**Comment by szha** [comment]
Job PR-882-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/16/index.html

**Comment by szha** [comment]
Job PR-882-17 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/17/index.html

**Comment by szha** [comment]
Job PR-882-19 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/19/index.html

**Comment by szha** [comment]
Job PR-882-21 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/21/index.html

File: vision/src/autogluon/vision/detector/detector.py
Comment (Innixma) [inline vision/src/autogluon/vision/detector/detector.py:308]: Should `pred_id` just be the index? Does it need to be its own column?

**Review by Innixma — COMMENTED** [review]
Once PR has been rebased I'll be happy to approve. I think some of the comments can be discussed after PR is merged.

File: vision/src/autogluon/vision/detector/detector.py
Comment (zhreshold) [reply vision/src/autogluon/vision/detector/detector.py:308]: this is the id of each predicted object

**Comment by szha** [comment]
Job PR-882-22 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-882/22/index.html

**Review by Innixma — APPROVED** [review]

**Pull request merged**
